[RAM] Add data model for scheduled and recurring snoozes

Zacqary opened the pull request:
## Summary

Closes #130724

Migrates the data model for rules from ~`muteAll`~/`snoozeEndTime` to ~`snoozeIndefinitely`~/`snoozeSchedule`.

~`snoozeIndefinitely` is a straight rename of `muteAll` and functions the same way.~ (Removed this as it's too large of a code change; `muteAll` is used in a ton of places)

`snoozeSchedule` is an array of scheduled snooze times, including options for recurring on a time interval or certain days of the week. Recurrence is formatted in the form of an [iCal RRULE](https://icalendar.org/iCalendar-RFC-5545/3-8-5-3-recurrence-rule.html), and this PR adds the [rrule package](https://www.npmjs.com/package/rrule) to parse these. Also adds [luxon](https://www.npmjs.com/package/luxon) which is its required peer dependency for dealing with time zones.

This just changes the data model and the task manager's parsing of it. APIs and UIs remain broadly unchanged for now.

The only change to the APIs and UIs are the addition of an `is_snoozed_until` property in the API response. This property is identical to the previously existing `snooze_end_time`, but it gets derived on the server side from the snooze schedule using the `getRuleSnoozeEndTime` function. **The reason we do this is because making the `rrule` package available in the frontend balloons the `alerting` plugin size, so all RRULE parsing needs to happen on the server. This may not last, as we're eventually going to need to generate RRULE strings from the snooze UI, but I don't want to include the decision to increase the alerting bundle size in this PR.**

**UPDATE 5/12**: To aid in filtering, `isSnoozedUntil` will now be stored in the SavedObject. This property is computed and added to the SavedObject either when the `_snooze` API is called (for snoozes that start as soon as they're added), OR when the task runner attempts to execute the rule (for recurring snoozes). As far as the task runner is concerned, the actual snooze source of truth will always be the `snoozeSchedule`. `isSnoozedUntil` will purely be for display and sorting.

**UPDATE 5/16**: Closes #132164 and uses `rRule` even for simple snoozes. When the user clicks "Snooze for 3 days" this will now set a snooze with an `rRule` of `{ dtstart: NOW, count: 1 }`

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

Zacqary commented on x-pack/plugins/alerting/common/is_rule_snoozed.test.ts:
```diff
@@ -0,0 +1,188 @@
+import sinon from 'sinon';
+import { isRuleSnoozed } from './is_rule_snoozed';
+
+const LOCAL_OFFSET = moment('2020-01-01T00:00:00.000').format('Z');
```
Appending the local offset is necessary for handling day of week recurrence.

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

Zacqary commented on package.json:
```diff
@@ -305,6 +306,7 @@
     "loader-utils": "^1.2.3",
     "lodash": "^4.17.21",
     "lru-cache": "^4.1.5",
+    "luxon": "^2.3.2",
```
This is a peer dependency of `rrule`

Zacqary commented on x-pack/plugins/alerting/server/saved_objects/migrations.ts:
```diff
@@ -859,6 +861,30 @@ function addMappedParams(
   return doc;
 }
 
+function convertSnoozes(
```
Note that this function doesn't do anything with the new RRULE support. It's just a simple migration of existing non-recurring snoozes into the new model.

Zacqary commented on x-pack/plugins/alerting/server/routes/find_rules.ts:
```diff
@@ -111,6 +111,9 @@ const rewriteBodyRes: RewriteResponseCase<FindResult<RuleTypeParams>> = ({
           connector_type_id: actionTypeId,
         })),
+        ...(snoozeSchedule != null
+          ? { is_snoozed_until: getRuleSnoozeEndTime({ snoozeSchedule, muteAll })?.toISOString() }
```
This is where we derive the current snooze end time from the snooze schedule, and handle RRULE parsing on the server side. This could be temporary if we decide to make the `rrule` package available on the frontend in a later PR.

XavierM commented on x-pack/plugins/alerting/common/rule_snooze_type.ts:
```diff
@@ -0,0 +1,14 @@
+  duration: number;
+  timeZone?: string;
+  id?: string;
+  rRule?: string;
```
Maybe we can use the type form the library `rrule` here.

Zacqary replied:
I don't want to import `rrule` outside of `/server` for now. That increases the alerting plugin bundle size to the point where it doesn't pass CI, so I'd prefer to wait to deal with that until we have a critical reason for including `rrule` on the frontend.

XavierM replied:
or we can just fork it for now

XavierM replied:
I meant that we can just copy the interface, so our API is well defined

kobelb replied:
The following came up when @XavierM and I were talking about timezones over Zoom. I think we should do what @XavierM just suggested where we don't support the complex `rRule` strings, and instead have a strongly typed object for `rRule` string which is equivalent to the type definition for the `rrule` library.

Zacqary replied:
This PR is already very large and duplicating the schema would be a lot of additional code, so I'm tracking this here: https://github.com/elastic/kibana/issues/132164

Zacqary commented on x-pack/plugins/alerting/server/saved_objects/migrations.ts:
```diff
@@ -859,6 +861,30 @@ function addMappedParams(
+            {
+              startTime: new Date().toISOString(),
+              duration: Date.parse(snoozeEndTime as string) - Date.now(),
+              timeZone: moment.tz.guess(),
```
I'm doing this actually in anticipation of showing the user-selected time zone in the UI. `startTime` is actually always represented in UTC (note the `.toISOString()`, which automatically converts to UTC), and `timeZone` is exclusively a signal to the UI about what timezone to display things in.

XavierM commented on x-pack/plugins/alerting/server/lib/is_rule_snoozed.ts:
```diff
@@ -0,0 +1,45 @@
+
+    // Check to see if now is during a recurrence of the snooze
+    if (rRule) {
+      const recurrenceRule = rrulestr(rRule);
```
do we need to worry about `rrulestr` to throw an exception if the rRule is not formatted correctly

XavierM commented on x-pack/plugins/alerting/server/saved_objects/mappings.ts:
```diff
@@ -185,9 +185,25 @@ export const alertMappings: SavedObjectsTypeMappingDefinition = {
-      type: 'date',
-      format: 'strict_date_time',
+    snoozeSchedule: {
+      properties: {
```
I do not think that it needs to be set nested to allow arrays. is there other reason to use nested?

XavierM reviewed (commented):
I am wondering if our `rrule` params in our API should be more than just a string and also follow this interface 
```

  freq: Frequency.YEARLY,
  dtstart: null,
  interval: 1,
  wkst: Days.MO,
  count: null,
  until: null,
  tzid: null,
  bysetpos: null,
  bymonth: null,
  bymonthday: null,
  bynmonthday: null,
  byyearday: null,
  byweekno: null,
  byweekday: null,
  bynweekday: null,
  byhour: null,
  byminute: null,
  bysecond: null,
  byeaster: null
}
``` 

Code looks good, I am going to run it locally

XavierM commented:
@arisonl brought a good requirement to use that if a user is scheduled a rule to snooze for every Monday 5pm, it remains at 5pm when we switch between daylight saving times. It also bring another point user watching this schedule snooze from different timezone.

I think we might have a little glitch with our solution right now, I think that we might need to store the timezone offset to make sure that our back end do the snooze at the right time. Maybe we also need to separate our time from the date to make the calculation easier on the server side to know when to snooze a rule.

To make it less confusing it on the UI, we will need to show the timezone pick for the snooze so users from Europe to East coast to West coast does not try to figure out at what time the rule is going to snooze. FYI @mdefazio

mdefazio commented:
> To make it less confusing it on the UI, we will need to show the timezone picked for the snooze, so our users from Europe to East coast to West coast does not try to figure out at what time the rule is going to snooze.

@XavierM @arisonl We can show the time zone in the date picker without issue. However, just to clarify, we want the time to show based on the user's Kibana settings, correct?  If someone in CET creates a schedule for Monday at 5pm.  And then a user in EST reviews that same schedule, it would show as Monday at 11am.  

But are we instead wanting to show 5pm CET (the timezone it was created in) and have the person in EST try and figure out what that is for them?

Again, we can show the UTC offsets easily within the UI, but i'm not sure I'm understanding the desired behavior. I would expect to see my current timezone and everything referencing that.

Zacqary commented:
> @arisonl brought a good requirement if a user is scheduling a rule to snooze for every Monday at 5pm, it should remains 5pm even of daylight saving times or not. It also bring another point to show this schedule snooze to users with a different timezone.

We can solve this by inputting the timezones as region names instead of offsets. `America/New_York` will always resolve to either EST or EDT depending on the date.

pmuellr commented:
> We can solve this by inputting the timezones as region names instead of offsets. America/New_York will always resolve to either EST or EDT depending on the date.

I was wondering if this was an option - specifying a timezone with a string independent of "daylight savings" (or other similar shifts like that).  Seems ideal.

XavierM commented:
> > @arisonl brought a good requirement if a user is scheduling a rule to snooze for every Monday at 5pm, it should remains 5pm even of daylight saving times or not. It also bring another point to show this schedule snooze to users with a different timezone.
> 
> We can solve this by inputting the timezones as region names instead of offsets. `America/New_York` will always resolve to either EST or EDT depending on the date.

@Zacqary, I think that will do the job too and might work better because kibana is already using timezones as region names in their advanced settings. we might need to do more work when kibana timezone is set to browser but not that much

Zacqary commented:
Updated the description with this but reposting here:

To aid in filtering, `isSnoozedUntil` will now be stored in the SavedObject. This property is computed and added to the SavedObject either when the `_snooze` API is called (for snoozes that start as soon as they're added), OR when the task runner attempts to execute the rule (for recurring snoozes). As far as the task runner is concerned, the actual snooze source of truth will always be the `snoozeSchedule`. `isSnoozedUntil` will purely be for display and sorting.

pmuellr commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -2142,8 +2146,18 @@ export class RulesClient {
+            snoozeSchedule: clearUnscheduledSnooze(attributes),
+          }
+        : {
+            snoozeSchedule: (attributes.snoozeSchedule ?? []).concat({
```
I think we'll want to do the `clearUnscheduledSnooze()` here as well?  If the user creates a duration-based snooze, and then creates another one, won't there then be two?  I guess that raises the question, is that a valid thing to do - have multiple duration-based snoozes without a real rRule schedule?  Seems like we should only have one, and it would be the last one the user set.  For example:

- time X, set snooze for 1 hour
- time X+1m, set snooze for 10 minutes

How long do we snooze?  X + 1h or X + 11m (X+1m + 10m snooze)?

pmuellr commented on x-pack/plugins/alerting/server/rules_client/rules_client.ts:
```diff
@@ -2817,3 +2860,9 @@ function parseDate(dateString: string | undefined, propertyName: string, default
   return parsedDate;
 }
+
+function clearUnscheduledSnooze(attributes: { snoozeSchedule?: RuleSnooze }) {
```
Not completely clear what this function is doing, but I think it's removing "mute all" snoozes that have a duration.

Zacqary replied:
Not exactly. This clears any "simple" snoozes, i.e. the user selects "Snooze for 3 days (starting now)" in the currently-existing UX, instead of e.g. "Snooze on Friday for 3 days" or "Snooze every Friday for 3 days" as made possible by the new functionality. It's the functional equivalent of setting `snoozeEndTime: null` in the previous version of the data model.

pmuellr reviewed (commented):
Added some comments.  It feels like we have too many combinations of properties that can be used to define "mute all forever", "mute all with a duration", and the eventual rRule-based snoozes.  Wondering if we can boil this down to a simpler set of properties, where we ALWAYS have an rRule in place.

I could totally believe we can't!  I just need to be convinced.  

And I think we need a bit of doc in the rule_snooze_type module describing these properties, as well as some doc in the PR describing all the combinations of things that can happen today, with an hint of how we plan on dealing with rRule schedules in the future.

Zacqary commented on x-pack/plugins/alerting/common/rule_snooze_type.ts:
```diff
@@ -0,0 +1,14 @@
+ * 2.0.
+ */
+
+export type RuleSnooze = Array<{
```
> Similar to this, it seems like we still have `muteAll` around, to handle the case of "mute forever". 

I tried merely changing the name of `muteAll` and it required changes in like 80 files. I agree we could probably remove it but not before feature freeze. This PR is huge enough already.

> Also not clear, but I don't think we're ever actually setting rRule right now?

No, this PR just changes the data model, not the APIs or the UIs. Setting the rRule comes when we actually implement the new recurring snooze functionality on the frontend.

> It seems like we could actually just create an rRule in this case, so wouldn't have to treat the snooze case as special with the startTime, and could then get rid of startTime completely. That way we'd always have the case where a snoozeSchedule entry would have a non-null rRule.

Possible, though having a `null` rRule is a quick and easy way to determine if a snooze is or is not recurring, without having to run any parsing on the rRule. I suppose re-implementing the `rrule` lib structure in the savedobject instead of using an rRule string could give us that for free, though, and just use `count: 1` as an indicator of a non-recurring snooze. Again, though, feature freeze is next week and we don't even have an API that sets scheduled and recurring snoozes yet.

Zacqary replied:
Tracking these changes in https://github.com/elastic/kibana/issues/132266

Zacqary replied:
Actually managed to implement moving `startTime` into the `rRule`, and requiring it for all snoozes. Still need to kick removing `muteAll` down the road to #132266 though.

pmuellr reviewed (approved):
LGTM; noted some things for follow ups - more doc (somewhere) and more robust parsing for the wacky day-of-week with optional numeric suffix thing (probably could be made into a grep pattern).

Zacqary commented on x-pack/plugins/alerting/server/lib/is_rule_snoozed.ts:
```diff
@@ -0,0 +1,67 @@
+  return byweekday.map((w) => {
+    if (typeof w === 'number') return w;
+    if (['SU', 'MO', 'TU', 'WE', 'TH', 'FR', 'SA'].includes(w)) return w as WeekdayStr;
+    const dayOfWeek = w.slice(-2) as WeekdayStr;
```
Yeah unfortunately the `rrule` library will parse an entire rrule string, but not just a segment of it, so we can't rely on its already robust parsing of `BYDAY`. I added a comment though.

pmuellr replied:
Not that we need to do it for this PR, but could we construct a legal but fake rRule with the BYDAY value, then send it to the rRule parser to just extract the relevant fields?

Zacqary replied:
Oh wait actually yeah that does work

Zacqary replied:
Oops nevermind had to use `origOptions`, this works fine
